source/common/config/watched_directory.cc

```diff
@@ -0,0 +1,43 @@
+#include "common/config/watched_directory.h"
+
+namespace Envoy {
+namespace Config {
+
+WatchedDirectory::WatchedDirectory(const envoy::config::core::v3::WatchedDirectory& config,
+                                   Event::Dispatcher& dispatcher, Filesystem::Instance& filesystem)
+    : filesystem_(filesystem),
+      literal_path_(filesystem.joinPath(config.root(), config.subdirectory())) {
+  resolvePath();
+  watcher_ = dispatcher.createFilesystemWatcher();
+  watcher_->addWatch(absl::StrCat(config.root(), "/"), Filesystem::Watcher::Events::MovedTo,
+                     [this](uint32_t) {
+                       resolvePath();
+                       cb_();
+                     });
+}
+
+void WatchedDirectory::resolveDataSourcePath(envoy::config::core::v3::DataSource& data_source) {
+  if (data_source.filename().empty() || resolved_path_.empty()) {
+    return;
+  }
+  try {
+    const auto result = filesystem_.splitPathFromFilename(data_source.filename());
+    if (result.directory_ == literal_path_) {
+      data_source.set_filename(filesystem_.joinPath(resolved_path_, result.file_));
```

Comment:
Since filename is set according to `resolved_path_`, does it mean that `data_source.filename()` will change each iteration according to realpath? Will `result.directory_ == literal_path_` still evaluate true the next round?

I was wondering: is the `subdirectory` parameter necessary at all in `WatchedDirectory`? Would it be reasonable to make watched directory completely independent concept from `DataSource.filename`?  That is, filename would be used for reading files, watched directory for the watch?

Reply:
Updated, I think both concerns are addressed.